### PR TITLE
[SV] Generate case support

### PR DESCRIPTION
**Pull request by teqdruid**
First step in #3230. Doesn't deal with any subtilities like `InnerRef` yet. Just the basic IR model and `ExportVerilog` support.

### CODE REVIEW

**Comment by teqdruid** [comment]
@darthscsi @seldridge This is ready for review.

File: lib/Dialect/SV/SVOps.cpp [hunk 1]
```diff
@@ -1648,6 +1648,72 @@ void CoverConcurrentOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+  }
+  p.printNewline();
+}
+
+LogicalResult GenerateCaseOp::verify() {
+  size_t numPatterns = casePatterns().size();
+  if (caseRegions().size() != numPatterns || caseNames().size() != numPatterns)
+    return emitOpError(
+        "Size of caseRegions, patterns, and caseNames must match");
+
```
Comment (darthscsi) [inline lib/Dialect/SV/SVOps.cpp:1700]: Do the types of the attributes have to match across case statements and do they have to match the condition?

File: test/Dialect/SV/generates.mlir [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+hw.module @Case1<NUM : i8> () -> () {
+  %fd = hw.constant 0x80000002 : i32
+  sv.generate "foo_case": {
+    sv.generate.case #hw.param.decl.ref<"NUM"> [
+      case (0, "case0") {
+        sv.initial {
+          sv.fwrite %fd, "case 0\n"
+        }
+      }
+      case (1 : i64, "case1") {
```
Comment (darthscsi) [inline test/Dialect/SV/generates.mlir:21]: Here, the 1 is an i64, but #NUM is an i8.

File: test/Dialect/SV/generates.mlir [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        hw.instance "print1" @PrintPath() -> ()
+      }
+      default ("dflt") {
```
Comment (darthscsi) [inline test/Dialect/SV/generates.mlir:24]: my comments about default are essentially, should this (at least internally, ignoring pretty printing) be:
`case (defaultCaseAttr, "dflt")`

Comment (darthscsi) [reply test/Dialect/SV/generates.mlir:24]: Or even a single case pattern with a null attr.

**Review by darthscsi — COMMENTED** [review]
It would be good to narrow the typing of attributes and verification.

**Comment by teqdruid** [comment]
@darthscsi I can't figure out how to resolve an Attribute's type to compare the cases against the condition. The 'evaluate` functions in HWAttributes.h require the instantiation parameters (right?) and `#hw.param.decl.ref`'s type is `none`. Am I missing something or is this functionality which needs to be added?

**Comment by teqdruid** [comment]
type comparisons rely on #3442.

**Pull request merged**